src/dali/plugin/input/rest/binance_com.py

```diff
@@ -0,0 +1,1295 @@
+
+                # If the dict already exists add another key, if not add new dict
+                if current_subscriptions.get(subscription[_ASSET]):
+                    current_subscriptions[subscription[_ASSET]][f"{RP2Decimal(subscription[_AMOUNT]):.13f}"] = subscription
+                else:
+                    current_subscriptions[subscription[_ASSET]] = {f"{RP2Decimal(subscription[_AMOUNT]):.13f}": subscription}
+
+            locked_redemptions = self.client.sapi_get_staking_stakingrecord(
+                params=({_START_TIME: current_start, _END_TIME: current_end, _PRODUCT: _STAKING, _TXN_TYPE: _REDEMPTION, _SIZE: _INTEREST_SIZE_LIMIT})
+            )
+            # [
+            #         {
+            #             "positionId": "12345",
+            #             "time": "1624147893000"
+            #             "asset": "BTC",
+            #             "amount": "1",
+            #             "deliverDate": "1624147895000"
+            #             "status": "PAID",
+            #         },
+            #         {
+            #             "positionId": "12346",
+            #             "time": "1624147993000"
+            #             "asset": "BTC",
+            #             "amount": "0.95",
+            #             "deliverDate": "1624148093000"
+            #             "status": "PAID",
+            #         }
+            # ]
+            # NOTE: all values are str
+
+            for redemption in locked_redemptions:
+
+                redemption_amount: str = f"{RP2Decimal(redemption[_AMOUNT]):.13f}"
+
+                # Check if there is a subscription with this asset and if the redemption amount is equal to the subscription amount
+                if redemption[_ASSET] in current_subscriptions and redemption_amount not in current_subscriptions[redemption[_ASSET]]:
+
+                    # If they do not equal we need to calculate what the amended principal should be based on total interest paid to that productId
+                    total_interest_earned: RP2Decimal = total_current_interest[redemption[_POSITION_ID]]
+                    original_principal: str = f"{(RP2Decimal(redemption[_AMOUNT]) + RP2Decimal(str(total_interest_earned))):.13f}"
+                    earliest_redemption_timestamp: int = 0
+
+                    if str(original_principal) in current_subscriptions[redemption[_ASSET]]:
+
+                        subscription_time: int = int(current_subscriptions[redemption[_ASSET]][str(original_principal)][_TIME])
+                        lockperiod_in_ms: int = int(current_subscriptions[redemption[_ASSET]][str(original_principal)][_LOCK_PERIOD]) * _ONE_DAY_IN_MS
+                        earliest_redemption_timestamp = subscription_time + lockperiod_in_ms
+
+                    else:
+                        self.__logger.error(
+                            "Internal Error: Principal (%s) minus paid interest (%s) does not equal returned principal"
```

Comment:
An internal error identifies an inconsistent/non-sane internal state that makes it impossible to proceed with execution: in such cases we typically raise an Exception but here it's an error log.

I think there are two options:

* if it's a normal error that doesn't reflect a non-sane state and doesn't hamper further execution, just remove "Internal Error" from the message string;
* if it is an internal error, throw an Exception.

Check https://github.com/eprbell/dali-rp2/blob/main/src/dali/plugin/input/rest/coinbase.py#L631 between lines 631 and 637 for an example of both cases.